README.md
# Tetris
just for fun


CHANGELOG.md
# changelog

## 1.1.0 (2023-05-02)
- random tetromino color by png
## 1.0.0 (2023-04-29)

- FULFILL MOST FUNCTION
- ADD SCORE
- ADD NEXT TETRIS


main.py
from settings import *
from tetris import Tetris, Text
import sys
import pathlib

class App:
    def __init__(self):
        pg.init()
        pg.display.set_caption("Tetris")
        self.screen = pg.display.set_mode(WIN_RES)
        self.clock = pg.time.Clock()
        self.set_timer()
        self.images = self.load_images()
        self.tetris = Tetris(self)
        self.text = Text(self)

    def load_images(self):
        files = [item for item in pathlib.Path(SPRITE_DIR_PATH).rglob('*.png') if item.is_file()]
        images = [pg.image.load(file).convert_alpha() for file in files]
        images = [pg.transform.scale(image, (TILE_SIZE, TILE_SIZE)) for image in images]
        return images

    def set_timer(self):
        self.user_event = pg.USEREVENT + 0
        self.fast_user_event = pg.USEREVENT + 1
        self.anim_trigger = False
        self.fast_user_event = False
        pg.time.set_timer(self.user_event, ANIM_TIME_INTERVAL)
        pg.time.set_timer(self.fast_user_event, FAST_ANIM_TIME_INTERVAL)

    def update(self):
        self.tetris.update()
        self.clock.tick(FPS)
    
    def draw(self):
        self.screen.fill(color=BG_COLOR) # 填入背景色
        self.screen.fill(color=FIELD_COLOR, rect=(0, 0, *FIELD_RES))
        self.tetris.draw() # 畫格線
        self.text.draw()
        pg.display.flip()

    def check_events(self):
        self.anim_trigger = False # 設成 False，使落下速度不影響
        self.fast_anim_trigger = False
        for event in pg.event.get():
            if event.type == pg.QUIT or (event.type == pg.KEYDOWN and event.key == pg.K_ESCAPE):
                pg.quit()
                sys.exit()
            elif event.type == pg.KEYDOWN:
                self.tetris.control(pressed_key=event.key)
            elif event.type == self.user_event:
                self.anim_trigger = True
            elif event.type == self.fast_user_event:
                self.fast_anim_trigger = True

    def run(self):
        while True:
           self.check_events()
           self.update()
           self.draw()

if __name__ == '__main__':
    app = App()
    app.run() 

Tetris.py
from settings import *
import math
from tetromino import Tetromino
import pygame.freetype as ft

class Text:
    def __init__(self, app):
        self.app = app
        self.font = ft.Font(FONT_PATH)
    
    def draw(self):
        self.font.render_to(self.app.screen, (WIN_W * 0.605, WIN_H * 0.02),
                            text='TETRIS', fgcolor='white',
                            size=TILE_SIZE * 1.65, bgcolor='black')
        self.font.render_to(self.app.screen, (WIN_W * 0.68, WIN_H * 0.22),
                            text='next', fgcolor='orange',
                            size=TILE_SIZE * 1.4, bgcolor='black')
        self.font.render_to(self.app.screen, (WIN_W * 0.64, WIN_H * 0.67),
                            text='score', fgcolor='orange',
                            size=TILE_SIZE * 1.4, bgcolor='black')
        self.font.render_to(self.app.screen, (WIN_W * 0.64, WIN_H * 0.8),
                            text=f'{self.app.tetris.score}', fgcolor='white',
                            size=TILE_SIZE * 1.8)

class Tetris:
    def __init__(self, app):
        self.app = app
        self.sprite_group = pg.sprite.Group()
        self.field_array = self.get_field_array()
        self.tetromino = Tetromino(self)
        self.next_tetromino = Tetromino(self, current=False)
        self.speed_up = False

        self.score = 0
        self.full_lines = 0
        self.points_per_lines = {0: 0, 1: 100, 2: 300, 3: 700, 4: 1500}
    
    def get_score(self):
        self.score += self.points_per_lines[self.full_lines]
        self.full_lines = 0
    
    def check_full_lines(self):
        row = FIELD_H -1
        for y in range(FIELD_H -1, -1, -1):
            for x in range(FIELD_W):
                self.field_array[row][x] = self.field_array[y][x]

                if self.field_array[y][x]:
                    self.field_array[row][x].pos = vec(x, y)

            if sum(map(bool, self.field_array[y])) < FIELD_W:
                row -= 1
            else:
                for x in range(FIELD_W):
                    self.field_array[row][x].alive = False
                    self.field_array[row][x] = 0

                self.full_lines += 1
    
    def put_tetromino_blocks_in_array(self):
        for block in self.tetromino.blocks:
            x, y = int(block.pos.x), int(block.pos.y)
            self.field_array[y][x] = block

    def get_field_array(self):
        return [[0 for x in range(FIELD_W)] for y in range(FIELD_H)]
    
    def is_game_over(self):
        if self.tetromino.blocks[0].pos.y == INIT_POS_OFFSET[1]:
            pg.time.wait(300)
            return True

    def check_tetromino_landing(self):
        if self.tetromino.landing:
            if self.is_game_over():
                self.__init__(self.app)
            else:
                self.speed_up = False
                self.put_tetromino_blocks_in_array()
                self.next_tetromino.current = True
                self.tetromino = self.next_tetromino
                self.next_tetromino = Tetromino(self, current=False)

    def control(self, pressed_key):
        if pressed_key == pg.K_LEFT:
            self.tetromino.move(direction='left')
        elif pressed_key == pg.K_RIGHT:
            self.tetromino.move(direction='right')
        elif pressed_key == pg.K_UP:
            self.tetromino.rotate()
        elif pressed_key == pg.K_SPACE:
            self.speed_up = True
            
    def draw_grid(self):
        for x in range (FIELD_W):
            for y in range(FIELD_H):
                # rect(參數, 隔線顏色, 長寬(left, top, width, height), 線的粗度)
                pg.draw.rect(self.app.screen, "black",
                             (x * TILE_SIZE, y * TILE_SIZE, TILE_SIZE, TILE_SIZE), 1)
    
    def update(self):
        trigger = [self.app.anim_trigger, self.app.fast_anim_trigger][self.speed_up]
        if trigger:
            self.check_full_lines()
            self.tetromino.update()
            self.check_tetromino_landing()
            self.get_score()
        self.sprite_group.update()
    
    def draw(self):
        self.draw_grid()
        self.sprite_group.draw(self.app.screen)

settings.py
import pygame as pg

vec = pg.math.Vector2

FPS = 60
FIELD_COLOR = (48, 39, 32) # 背景色
BG_COLOR = (24, 89, 117)

SPRITE_DIR_PATH = 'sprites'

FONT_PATH = 'font/SFPixelateShaded-Bold.ttf'

ANIM_TIME_INTERVAL = 200 # milliseconds 
FAST_ANIM_TIME_INTERVAL = 15

TILE_SIZE = 40 # 每格長度
FIELD_SIZE = FIELD_W, FIELD_H = 10, 20 # 長、寬幾格
FIELD_RES = FIELD_W * TILE_SIZE, FIELD_H * TILE_SIZE

FIELD_SCALE_W, FIELD_SCALE_H = 1.7, 1.0
WIN_RES = WIN_W, WIN_H = FIELD_RES[0] * FIELD_SCALE_W, FIELD_RES[1] * FIELD_SCALE_H

INIT_POS_OFFSET = vec(FIELD_W // 2 - 1, 0) #設定初始位置
NEXT_POS_OFFSET = vec(FIELD_W * 1.3, FIELD_H * 0.45)
MOVE_DIRECTIONS = {'left': vec(-1, 0), 'right': vec(1, 0), 'down': vec(0, 1)}

TETROMINO = {
    'T': [(0, 0), (-1, 0), (1, 0), (0, -1)],
    'O': [(0, 0), (0, -1), (1, 0), (1, -1)],
    'J': [(0, 0), (-1, 0), (0, -1), (0, -2)],
    'L': [(0, 0), (1, 0), (0, -1), (0, -2)],
    'I': [(0, 0), (0, 1), (0, -1), (0, -2)],
    'S': [(0, 0), (-1, 0), (0, -1), (1, -1)],
    'Z': [(0, 0), (1, 0), (0, -1), (-1, -1)]
}

COLORS = {
    'black': (0,0,0),
    'red': (255,0,0),
    'blue': (0,0,255),
    'green': (0,255,0),
    'yellow': (0, 255, 255),
    'orange': (255, 255, 0),
    'sliver': (192, 192, 192)
}

tetromino.py
from settings import *
import random

class Block(pg.sprite.Sprite):
    def __init__(self, tetromino, pos):
        self.tetromino = tetromino
        self.pos = vec(pos) + INIT_POS_OFFSET
        self.next_pos = vec(pos) + NEXT_POS_OFFSET
        self.alive = True

        super().__init__(tetromino.tetris.sprite_group)
        self.image = tetromino.image
        # self.image = pg.Surface([TILE_SIZE, TILE_SIZE]) 
        # pg.draw.rect(self.image, (0, 0, 255) , (1, 1, TILE_SIZE - 2, TILE_SIZE - 2), border_radius=8)
        # random.choice(list(COLORS.values()))
        self.rect = self.image.get_rect()

    def is_alive(self):
        if not self.alive:
            self.kill()

    def rotate(self, pivot_pos):
        translated = self.pos - pivot_pos
        rotated = translated.rotate(90)
        return rotated + pivot_pos
    
    def set_rect_pos(self):
        pos = [self.next_pos, self.pos][self.tetromino.current]
        self.rect.topleft = pos * TILE_SIZE

    def update(self):
        self.is_alive()
        self.set_rect_pos()

    def is_collide(self, pos): # 是否撞到邊界
        x, y = int(pos.x), int(pos.y)
        if 0 <= x < FIELD_W and y < FIELD_H and (
                y < 0 or not self.tetromino.tetris.field_array[y][x]):
            return False
        return True
    
class Tetromino:
    def __init__(self, tetris, current=True):
        self.tetris = tetris
        self.shape = random.choice(list(TETROMINO.keys()))
        self.image = random.choice(tetris.app.images)
        self.blocks = [Block(self, pos) for pos in TETROMINO[self.shape]]
        # self.color = random.choice(list(COLORS.values()))
        self.landing = False
        self.current = current
    
    def rotate(self):
        pivot_pos = self.blocks[0].pos
        new_block_positions = [block.rotate(pivot_pos) for block in self.blocks]

        if not self.is_collide(new_block_positions):
            for i, block in enumerate(self.blocks):
                block.pos = new_block_positions[i]
    
    def is_collide(self, block_positions): 
        return any(map(Block.is_collide, self.blocks, block_positions))
    
    def move(self, direction):
        move_direction = MOVE_DIRECTIONS[direction]
        new_block_positions = [block.pos + move_direction for block in self.blocks]
        is_collide = self.is_collide(new_block_positions)

        if not is_collide: 
            for block in self.blocks:
                block.pos += move_direction
        elif direction == 'down':
            self.landing = True

    def update(self):
        self.move(direction='down')

Tetris old ver.py
import pygame
import random

class Tetris:
    def __init__(self):
        pygame.init()
        self.width = 800
        self.height = 700
        self.screen = pygame.display.set_mode((self.width, self.height))
        pygame.display.set_caption("Tetris")
        self.clock = pygame.time.Clock()

        # Define the shapes of the Tetrominos
        self.shapes = [
            [[1, 1, 1],
             [0, 1, 0]],

            [[0, 2, 2],
             [2, 2, 0]],

            [[3, 3, 0],
             [0, 3, 3]],

            [[4, 0, 0],
             [4, 4, 4]],

            [[0, 0, 5],
             [5, 5, 5]],

            [[6, 6, 6, 6]],

            [[7, 7],
             [7, 7]]
        ]

        self.colors = [
            (0, 0, 0),
            (255, 0, 0),
            (0, 255, 0),
            (0, 0, 255),
            (255, 255, 0),
            (255, 0, 255),
            (0, 255, 255),
            (192, 192, 192)
        ]

        self.block_size = 30
        self.grid_width = 10
        self.grid_height = 20
        self.grid = [[0 for _ in range(self.grid_width)] for _ in range(self.grid_height)]
        self.score = 0
        self.active_tetromino = self.get_random_tetromino()
        self.next_tetromino = self.get_random_tetromino()
        self.tetromino_x = 3
        self.tetromino_y = 0

    def get_random_tetromino(self):
        return random.choice(self.shapes)

    def draw_block(self, x, y, color):
        pygame.draw.rect(self.screen, self.colors[color], pygame.Rect(x, y, self.block_size, self.block_size))
        pygame.draw.rect(self.screen, (255, 255, 255), pygame.Rect(x, y, self.block_size, self.block_size), 1)

    def draw_tetromino(self, x, y, tetromino):
        for i in range(len(tetromino)):
            for j in range(len(tetromino[i])):
                if tetromino[i][j] != 0:
                    self.draw_block((j + x) * self.block_size, (i + y) * self.block_size, tetromino[i][j])

    def draw_grid(self):
        for i in range(self.grid_height):
            for j in range(self.grid_width):
                if self.grid[i][j] != 0:
                    self.draw_block(j * self.block_size, i * self.block_size, self.grid[i][j])

    def move_tetromino(self, dx, dy):
        if not self.check_collision(dx, dy, self.active_tetromino):
            self.tetromino_x += dx
            self.tetromino_y += dy
            return True
        else:
            return False

    def rotate_tetromino(self):
        new_tetromino = []
        for i in range(len(self.active_tetromino[0])):
            new_row = []            
            for row in self.active_tetromino[::-1]:
                new_row.append(row[i])
            new_tetromino.append(new_row)
        if not self.check_collision(0, 0, new_tetromino):
            self.active_tetromino = new_tetromino

    def check_collision(self, dx, dy, tetromino):
        for i in range(len(tetromino)):
            for j in range(len(tetromino[i])):
                if tetromino[i][j] != 0:
                    x = self.tetromino_x + j + dx
                    y = self.tetromino_y + i + dy
                    if x < 0 or x >= self.grid_width or y >= self.grid_height or self.grid[y][x] != 0:
                        return True
        return False

    def fix_tetromino(self):
        for i in range(len(self.active_tetromino)):
            for j in range(len(self.active_tetromino[i])):
                if self.active_tetromino[i][j] != 0:
                    x = self.tetromino_x + j
                    y = self.tetromino_y + i
                    self.grid[y][x] = self.active_tetromino[i][j]
        self.clear_rows()
        self.active_tetromino = self.next_tetromino
        self.next_tetromino = self.get_random_tetromino()
        self.tetromino_x = 3
        self.tetromino_y = 0
        if self.check_collision(0, 0, self.active_tetromino):
            self.game_over()

    def clear_rows(self):
        num_rows_cleared = 0
        for i in range(len(self.grid)):
            if all(self.grid[i]):
                self.grid.pop(i)
                self.grid.insert(0, [0 for _ in range(self.grid_width)])
                num_rows_cleared += 1
        if num_rows_cleared > 0:
            self.score += 10 * (2 ** (num_rows_cleared - 1))

    def game_over(self):
        font = pygame.font.SysFont(None, 100)
        text = font.render("Game Over", True, (255, 0, 0))
        self.screen.blit(text, (self.width // 2 - text.get_width() // 2, self.height // 2 - text.get_height() // 2))
        pygame.display.flip()
        pygame.time.wait(3000)
    
    def display_score(self):
        font = pygame.font.SysFont(None, 24)
        text = font.render("Score: {}".format(self.score), True, (255,255,255))
        self.screen.blit(text, (10,10))
        pygame.display.flip()

    def run(self):
        while True:
            for event in pygame.event.get():
                if event.type == pygame.QUIT:
                    pygame.quit()
                    quit()
                elif event.type == pygame.KEYDOWN:
                    if event.key == pygame.K_LEFT:
                        self.move_tetromino(-1, 0)
                    elif event.key == pygame.K_RIGHT:
                        self.move_tetromino(1, 0)
                    elif event.key == pygame.K_DOWN:
                        self.move_tetromino(0, 1)
                    elif event.key == pygame.K_UP:
                        self.rotate_tetromino()

            self.screen.fill((255,255,255))
            self.draw_grid()
            self.draw_tetromino(self.tetromino_x, self.tetromino_y, self.active_tetromino)
            pygame.display.flip()

            if not self.move_tetromino(0, 1):
                self.fix_tetromino()

            self.clock.tick(5)

if __name__ == '__main__':
    game = Tetris()
    game.run()
    pygame.quit()

